### ISSUE TITLE
Newbie question regarding validation

### ISSUE THREAD

**Issue by hmihail** [canton7/Stylet#29]
I am a beginner in MVVM and I am not sure which is the best approach:

I have some domain-level models, like `User`, `Product`, `ProductGroup`, etc.
My ViewModels only have one one property which implements INPC, like this:

```
public ProductGroupViewModel(ProductGroup productGroup) : Screen
        {
            _productGroup = productGroup;
        }

        private ProductGroup _productGroup;
        public ProductGroup ProductGroup
        {
            get => _productGroup;
            set
            {
                _productGroup = value;
                SetAndNotify(ref _productGroup, value);
            }
        }
}
```

In the View, all the binding is done like this, and this works:
`Text="{Binding ProductGroup.Name}"`

I also use the same View / ViewModel to create a new `ProductGroup` and edit an existing one, through the contructor. But in this way, I cannot use Stylet IoC to inject the validator, because my contructing viewmodel does not know how to resolve such a constructor:

```
public ProductGroupViewModel(ProductGroup productGroup, IModelValidator<ProductGroupViewModel> validator) : base(validator)
        { ... }
```
What would be the recommended way to solve this issue?
Thanks in advance.

**Comment by canton7** [comment]
There are two ways to solve this, based on how lazy you're feeling.

The first is to change the constructor of `ProductGroupViewModel` so it only takes the `IModelValidator<ProductGroupViewModel>`. Then inject a `Func<ProductGroupViewModel>` into the "constructing viewmodel", and do

```csharp
var vm = productGroupViewModelFactory();
vm.ProductGroup = productGroup;
```

The second is to define a `ProductGroupViewModelFactory`:

```csharp
public interface IProductGroupViewModelFactory
{
    ProductGroupViewModel Create(ProductGroup productGroup);
}
public class ProductGroupViewModelFactory : IProductGroupViewModelFactory
{
    private readonly IModelValidator<ProductGroupViewModel> validator;

    public ProductGroupViewModelFactory(IModelValidator<ProductGroupViewModel> validator)
    {
        this.validator = validator;
    }
    
    public ProductGroupViewModel Create(ProductGroup productGroup)
    {
        return new ProductGroupViewModel(productGroup, this.validator);
    }
}
```

Then inject the `IProductGroupViewModelFactory` into the constructing viewmodel.

Some IoC containers can create that ProductGroupViewModelFactory implementation for you - StyletIoC isn't smart enough.

**Comment by hmihail** [comment]
Thanks for the reply.
I want to make it the cleaner (second) approach, but since I have more that 10 models, there will be some classes to write. Also, I need to register all into the IoC, right?

Thanks!

**Comment by canton7** [comment]
Yes and yes. It might be worth grabbing a more powerful IoC container which can generate those classes for you (e.g. [Autofac](http://docs.autofac.org/en/latest/advanced/delegate-factories.html)).

**Comment by hmihail** [comment]
Thanks for your support!

**Comment by hmihail** [comment]
I think I am lost again.

I've created the `ProductGroupViewModelfactory` as you suggested, and injected it into the contructor of `ProductGroupViewModel`.

My problem is now this: I have the `ShellViewModel` (main form) which calls `ProductGroupsView` to display all `ProductGroups`. On this view is a button to add a new `ProductGroup` which constructs the view using the factory. But from `ShellView` I inject the `WindowManager` into the `ProductGroupsView`, so all my constructors back to `ShellView` need to be altered.

Am I missing something?

**Comment by canton7** [comment]
Wait, did you mean `ShellViewModel` and `ProductGroupsViewModel` instead of the views?

**Comment by canton7** [comment]
Could you also post your constructor signatures - it's hard to follow your description.

**Comment by hmihail** [comment]
Here they are:

    public ShellViewModel(IWindowManager windowManager)
    {
        this.windowManager = windowManager;
    }
    
I open the view for ProductGroups:
-> Here the contructor is wrong

    var productGroupsViewmodel = new ProductGroupsViewModel(windowManager);
    windowManager.ShowDialog(productGroupsViewmodel);


    public ProductGroupsViewModel(IWindowManager windowManager, IProductGroupViewModelFactory factory)
    {
        ...
    }
    
And later on I open the ProductGroup view:

    var productGroup = new ProductGroup();
    var productGroupViewModel = modelFactory.Create(productGroup);
    bool? result = windowManager.ShowDialog(productGroupViewModel);

**Comment by canton7** [comment]
Inject a `Func<ProductGroupsViewModel>` into the `ShellViewModel` (or write an explicit factory if you'd rather). Then you do this:

```csharp
public class ShellViewModel
{
    private readonly IWindowManager windowManager;
    private readonly Func<ProductGroupsViewModel> productGroupsViewModelFactory;

    public ShellViewModel(IWindowManager windowManager, Func<ProductGroupsViewModel> productGroupsViewModelFactory)
    {
        this.windowManager = windowManager;
        this.productGroupsViewModelFactory = productGroupsViewModelFactory;
    }

    public void Foo()
    {
        this.windowManager.ShowDialog(this.productGroupsViewModelFactory());
    }
}
```

Generally in DI you avoid instantiating classes yourself. Instead, your constructor accepts an instance of the class (if you just need one instance), or a factory which can create instances of the class. That way, you don't need to know how to construct an instance of the class, and so the class can have its own dependencies injected into it without affecting you.

**Comment by hmihail** [comment]
Yes, that chain looked ugly. However, I don't know how to register that `Func` to the IoC, can you please provide an example?

Thanks alot!

**Comment by canton7** [comment]
You don't need to. [See the docs](https://github.com/canton7/Stylet/wiki/StyletIoC-Factories#the-solution).

**Comment by hmihail** [comment]
It looks like I will need to use another IoC container, because me having more than 10 `ViewModels` needed in `ShellViewModel`, I will have to inject 10 `Func` into it, right?

**Comment by canton7** [comment]
Using another IoC container won't solve that issue. Yes, if you have a separate factory for each ViewModel, then ShellViewModel will need 10 factories injected into it.

There are a couple of other things you could do to reduce this pain:

1. Inject the container itself (or a wrapper around it), then get new instances with `container.Get<ProductGroupsViewModel>()`. This is the service locator pattern.
2. Write a single custom factory which can generate `ProductGroupsViewModel`s, and all of the other ViewModel types which the `ShellViewModel` needs to create.

**Comment by hmihail** [comment]
For the moment I ended up injecting the container itself, just to see how it works. My problem is now that my view starts as it should, with the Save button disabled. But when I fill the only field that needs to be filled, the state of the Save button does't change, it stays disabled.

Did I miss something? I followed the example in the source code and I think is the same logic.

Here is a part of the ViewModel:

    public ProductGroupViewModel(ProductGroup productGroup, IModelValidator<ProductGroupViewModel> validator) : base(validator)
    {
        _productGroup = productGroup;

        this.Validate();
    }

    protected override void OnValidationStateChanged(IEnumerable<string> changedProperties)
    {
        base.OnValidationStateChanged(changedProperties);
    }
    
    public bool CanSaveItem
    {
        get { return !this.HasErrors; }
    }

    public void SaveItem()
    {
        if (this.Validate())
        {
            this.RequestClose(true);
        }
    }

**Comment by canton7** [comment]
You example doesn't include that field, so I can't tell. Please create a minimal sample which reproduces the issue, then post both the C# and the XAML.

**Comment by hmihail** [comment]
I addded this to the `OnValidationStateChanged`:

    this.NotifyOfPropertyChange(() => this.CanSubmit);

From what I can see there is no way to revalidate the form when that field is changed, so in case of validation succes, the button is enabled.

my XAML code is this:

    <TextBox Text="{Binding ProductGroup.Name, UpdateSourceTrigger=PropertyChanged}"></TextBox>

**Comment by canton7** [comment]
> From what I can see there is no way to revalidate the form when that field is changed, so in case of validation succes, the button is enabled.

I'm not sure what you're trying to say there.

> Binding ProductGroup.Name

I need more information to work with. That's why I asked for all of the code in a reduced sample, not one-line snippets with no context.

What's the DataContext here? Is it the ProductGroupViewModel?

**Comment by hmihail** [comment]
Yes, the data context is `ProductGroupViewModel`

**Comment by canton7** [comment]
If you have `Binding ProductGroup.Name`, then whatever `ProductGroup` is needs to be the one validating its `Name` property. 

Only the Name property is changing. The ProductGroup property is not changing.

**Comment by hmihail** [comment]
Here is a small sample with the issue. Thanks again for your help!

[StyletValidation.zip](https://github.com/canton7/Stylet/files/1028960/StyletValidation.zip)

**Comment by canton7** [comment]
The issue is as I described above.

**Comment by hmihail** [comment]
That one I was trying to avoid. This is a simple example, but my classes will have many properties. I feel that there is a lot of duplication, creating all that fields pairs, one in the domain model and one in the view model?

**Comment by canton7** [comment]
That is how MVVM works.

In practice there are often differences between the ViewModel and the model. Some properties may be grouped in the Model but separate in the ViewModel. Maybe the ViewModel uses a ComboBox, where the Model uses an enum or a string. Sometimes the ViewModel uses a text input which is a string, but the Model uses an integer or a double or a DateTime. The separation gives you wriggle room.

**Comment by hmihail** [comment]
So in my ViewModel I will always have to initialize properties from the Model, and in the Save method I have to set them back so I can persist the Model in database?

**Comment by canton7** [comment]
Yes

**Comment by hmihail** [comment]
Is it than recommended to use something like Fody to reduce the ammount of code required?

**Comment by canton7** [comment]
I use Fody to reduce the PropertyChanged boilerplate, yes, as I say in the wiki somewhere.

**Comment by hmihail** [comment]
Thanks alot! You have been a great help!
